lib/charms/smtp_integrator/v0/smtp.py

```diff
@@ -0,0 +1,286 @@
+        password: The SMTP AUTH password to use for the outgoing SMTP relay.
+        auth_type: The type used to authenticate with the SMTP relay.
+        transport_security: The security protocol to use for the outgoing SMTP relay.
+        domain: The domain used by the sent emails from SMTP relay.
+    """
+
+    @property
+    def host(self) -> str:
+        """Fetch the SMTP host from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("host")
+
+    @property
+    def port(self) -> int:
+        """Fetch the SMTP port from the relation."""
+        assert self.relation.app
+        return int(self.relation.data[self.relation.app].get("port"))
+
+    @property
+    def user(self) -> str:
+        """Fetch the SMTP user from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("user")
+
+    @property
+    def password(self) -> str:
+        """Fetch the SMTP password from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("password")
+
+    @property
+    def auth_type(self) -> str:
+        """Fetch the SMTP auth type from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("auth_type")
+
+    @property
+    def transport_security(self) -> str:
+        """Fetch the SMTP transport security protocol from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("transport_security")
+
+    @property
+    def domain(self) -> str:
+        """Fetch the SMTP domain from the relation."""
+        assert self.relation.app
+        return self.relation.data[self.relation.app].get("domain")
+
+
+class SmtpRequiresEvents(ops.CharmEvents):
+    """SMTP events.
+
+    This class defines the events that a SMTP requirer can emit.
+
+    Attrs:
+        smtp_data_available: the SmtpDataAvailableEvent.
+    """
+
+    smtp_data_available = ops.EventSource(SmtpDataAvailableEvent)
+
+
+class SmtpRequires(ops.Object):
+    """Requirer side of the SMTP relation.
+
+    Attrs:
+        on: events the provider can emit.
+    """
+
+    on = SmtpRequiresEvents()
+
+    def __init__(self, charm: ops.CharmBase, relation_name: str = DEFAULT_RELATION_NAME) -> None:
+        """Construct.
+
+        Args:
+            charm: the provider charm.
+            relation_name: the relation name.
+        """
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+        self.framework.observe(charm.on[relation_name].relation_changed, self._on_relation_changed)
+
+    def _on_relation_changed(self, event: ops.RelationChangedEvent) -> None:
+        """Event emitted when the relation has changed.
+
+        Args:
+            event: event triggering this handler.
+        """
+        assert event.relation.app
```

Comment:
Here, and anywhere else that has an `assert`, it is generally more user friendly to raise an exception rather than a generic `AssertionError` since the custom exception can be specifically handled whereas a general `AssertionError` must be generically handled